gateway/ln-gateway/src/ng/tests.rs

```diff
@@ -583,3 +585,94 @@ async fn test_gateway_cannot_pay_expired_invoice() -> anyhow::Result<()> {
+        return Ok(());
+    }
+
+    let fixtures = fixtures();
+    let lnd = fixtures.lnd().await;
+    let cln = fixtures.cln().await;
+
+    let GetNodeInfoResponse { pub_key, alias: _ } = lnd.info().await?;
+    let lnd_public_key = PublicKey::from_slice(&pub_key)?;
+
+    let GetNodeInfoResponse { pub_key, alias: _ } = cln.info().await?;
+    let cln_public_key = PublicKey::from_slice(&pub_key)?;
+    let all_keys = vec![lnd_public_key, cln_public_key];
+
+    for gateway_type in vec![LightningNodeName::Cln, LightningNodeName::Lnd] {
+        for num_route_hints in 0..=1 {
+            let gateway_ln = match gateway_type {
+                LightningNodeName::Cln => fixtures.cln().await,
+                LightningNodeName::Lnd => fixtures.lnd().await,
+            };
+
+            let GetNodeInfoResponse { pub_key, alias: _ } = gateway_ln.info().await?;
+            let public_key = PublicKey::from_slice(&pub_key)?;
+
+            tracing::info!("Creating federation with gateway type {gateway_type}. Number of route hints: {num_route_hints}");
+
+            let fed = fixtures.new_fed().await;
+            let user_client = fed.new_client().await;
+            let mut gateway = fixtures.new_gateway(gateway_ln, num_route_hints).await;
+            gateway.connect_fed(&fed).await;
+
+            let invoice_amount = sats(100);
+            let (_invoice_op, invoice) = user_client
+                .create_bolt11_invoice(invoice_amount, "description".into(), None)
+                .await?;
+            let route_hints = invoice.route_hints();
+
+            match num_route_hints {
+                0 => {
+                    // If there's no additional route hints, we're expecting a single route hint
+                    // with a single hop on the invoice, where the hop is the
+                    // public key of the gateway lightning node
+                    assert_eq!(
+                        route_hints.len(),
+                        1,
+                        "Found {} route hints when 1 was expected for {gateway_type} gateway",
+                        route_hints.len()
+                    );
+                    let route_hint = route_hints.get(0).unwrap();
+                    assert_eq!(
+                        route_hint.0.len(),
+                        1,
+                        "Found {} hops when 1 was expected for {gateway_type} gateway",
+                        route_hint.0.len()
+                    );
+                    let route_hint_pub_key = route_hint.0.get(0).unwrap().src_node_id;
+                    assert_eq!(
+                        route_hint_pub_key, public_key,
+                        "Public key of route hint hop did not match expected public key"
+                    );
+                }
+                _ => {
+                    // If there's more than one route hint, we're expecting the invoice to contain
+                    // `num_route_hints`. Each should have 2 hops.
```

Comment:
We might want to change that to also including the one-hop route hint even when including the 2-hop ones. But that can be a separate PR.